{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n",
      "\n",
      "real\t2m16.646s\n",
      "user\t0m12.185s\n",
      "sys\t0m4.622s\n"
     ]
    }
   ],
   "source": [
    "time conda create --name deeptools_kernel --file deeptools_kernel.yml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "&nbsp;\n",
    "\n",
    "Now use `conda activate` to load the environment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "conda activate deeptools_kernel"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "&nbsp;\n",
    "\n",
    "Now use `conda list` to see the packages it installed (a long list):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# packages in environment at /projects/.colostate.edu/dcking/software/anaconda/envs/deeptools_kernel:\n",
      "#\n",
      "# Name                    Version                   Build  Channel\n",
      "_libgcc_mutex             0.1                 conda_forge    conda-forge\n",
      "_openmp_mutex             4.5                       2_gnu    conda-forge\n",
      "anyio                     3.5.0            py38h06a4308_0  \n",
      "appdirs                   1.4.4              pyhd3eb1b0_0  \n",
      "argon2-cffi               21.3.0             pyhd3eb1b0_0  \n",
      "argon2-cffi-bindings      21.2.0           py38h7f8727e_0  \n",
      "asttokens                 2.0.5              pyhd3eb1b0_0  \n",
      "async-lru                 2.0.4            py38h06a4308_0  \n",
      "attrs                     23.1.0           py38h06a4308_0  \n",
      "babel                     2.11.0           py38h06a4308_0  \n",
      "backcall                  0.2.0              pyhd3eb1b0_0  \n",
      "bash                      5.0.018              h14c3975_0  \n",
      "bash_kernel               0.9.3              pyh4f82c71_0    conda-forge\n",
      "beautifulsoup4            4.12.2           py38h06a4308_0  \n",
      "bleach                    4.1.0              pyhd3eb1b0_0  \n",
      "brotli-python             1.0.9            py38h6a678d5_7  \n",
      "bzip2                     1.0.8                h7b6447c_0  \n",
      "c-ares                    1.19.1               h5eee18b_0  \n",
      "ca-certificates           2023.08.22           h06a4308_0  \n",
      "certifi                   2023.11.17       py38h06a4308_0  \n",
      "cffi                      1.16.0           py38h5eee18b_0  \n",
      "charset-normalizer        2.0.4              pyhd3eb1b0_0  \n",
      "comm                      0.1.2            py38h06a4308_0  \n",
      "cryptography              41.0.3           py38hdda0065_0  \n",
      "cycler                    0.11.0             pyhd3eb1b0_0  \n",
      "debugpy                   1.6.7            py38h6a678d5_0  \n",
      "decorator                 5.1.1              pyhd3eb1b0_0  \n",
      "deeptools                 3.5.1              pyhdfd78af_1    bioconda\n",
      "deeptoolsintervals        0.1.9            py38h4c6a040_6    bioconda\n",
      "defusedxml                0.7.1              pyhd3eb1b0_0  \n",
      "entrypoints               0.4              py38h06a4308_0  \n",
      "executing                 0.8.3              pyhd3eb1b0_0  \n",
      "freetype                  2.12.1               h4a9f257_0  \n",
      "icu                       58.2                 he6710b0_3  \n",
      "idna                      3.4              py38h06a4308_0  \n",
      "importlib-metadata        6.0.0            py38h06a4308_0  \n",
      "importlib_metadata        6.0.0                hd3eb1b0_0  \n",
      "importlib_resources       6.1.0            py38h06a4308_0  \n",
      "ipykernel                 6.25.0           py38h2f386ee_0  \n",
      "ipython                   8.12.2           py38h06a4308_0  \n",
      "ipython_genutils          0.2.0              pyhd3eb1b0_1  \n",
      "jedi                      0.18.1           py38h06a4308_1  \n",
      "jinja2                    3.1.2            py38h06a4308_0  \n",
      "json5                     0.9.6              pyhd3eb1b0_0  \n",
      "jsonschema                4.19.2           py38h06a4308_0  \n",
      "jsonschema-specifications 2023.7.1         py38h06a4308_0  \n",
      "jupyter-lsp               2.2.0            py38h06a4308_0  \n",
      "jupyter_client            8.6.0            py38h06a4308_0  \n",
      "jupyter_core              5.5.0            py38h06a4308_0  \n",
      "jupyter_events            0.8.0            py38h06a4308_0  \n",
      "jupyter_server            1.24.0             pyhd8ed1ab_0    conda-forge\n",
      "jupyter_server_terminals  0.4.4            py38h06a4308_1  \n",
      "jupyterlab                3.3.2              pyhd3eb1b0_0  \n",
      "jupyterlab_pygments       0.1.2                      py_0  \n",
      "jupyterlab_server         2.25.1           py38h06a4308_0  \n",
      "kiwisolver                1.4.4            py38h6a678d5_0  \n",
      "krb5                      1.20.1               h143b758_1  \n",
      "ld_impl_linux-64          2.38                 h1181459_1  \n",
      "libblas                   3.9.0           16_linux64_openblas    conda-forge\n",
      "libcblas                  3.9.0           16_linux64_openblas    conda-forge\n",
      "libcurl                   8.4.0                h251f7ec_1  \n",
      "libdeflate                1.18                 h0b41bf4_0    conda-forge\n",
      "libedit                   3.1.20221030         h5eee18b_0  \n",
      "libev                     4.33                 h7f8727e_1  \n",
      "libffi                    3.4.4                h6a678d5_0  \n",
      "libgcc-ng                 13.2.0               h807b86a_3    conda-forge\n",
      "libgfortran-ng            11.2.0               h00389a5_1  \n",
      "libgfortran5              11.2.0               h1234567_1  \n",
      "libgomp                   13.2.0               h807b86a_3    conda-forge\n",
      "liblapack                 3.9.0           16_linux64_openblas    conda-forge\n",
      "libnghttp2                1.57.0               h2d74bed_0  \n",
      "libopenblas               0.3.21               h043d6bf_0  \n",
      "libpng                    1.6.39               h5eee18b_0  \n",
      "libsodium                 1.0.18               h7b6447c_0  \n",
      "libssh2                   1.10.0               hdbd6064_2  \n",
      "libstdcxx-ng              11.2.0               h1234567_1  \n",
      "libzlib                   1.2.13               hd590300_5    conda-forge\n",
      "markupsafe                2.1.1            py38h7f8727e_0  \n",
      "matplotlib-base           3.1.3            py38hef1b27d_0  \n",
      "matplotlib-inline         0.1.6            py38h06a4308_0  \n",
      "mistune                   2.0.4            py38h06a4308_0  \n",
      "nbclassic                 0.5.6              pyh8b2e9e2_0    conda-forge\n",
      "nbclient                  0.8.0            py38h06a4308_0  \n",
      "nbconvert-core            7.7.1              pyhd8ed1ab_0    conda-forge\n",
      "nbconvert-webpdf          7.7.1              pyhd8ed1ab_0    conda-forge\n",
      "nbformat                  5.9.2            py38h06a4308_0  \n",
      "ncurses                   6.4                  h6a678d5_0  \n",
      "nest-asyncio              1.5.6            py38h06a4308_0  \n",
      "notebook-shim             0.2.3            py38h06a4308_0  \n",
      "numpy                     1.22.3           py38h99721a1_2    conda-forge\n",
      "openssl                   3.2.0                hd590300_1    conda-forge\n",
      "overrides                 7.4.0            py38h06a4308_0  \n",
      "packaging                 23.1             py38h06a4308_0  \n",
      "pandocfilters             1.5.0              pyhd3eb1b0_0  \n",
      "parso                     0.8.3              pyhd3eb1b0_0  \n",
      "pexpect                   4.8.0              pyhd3eb1b0_3  \n",
      "pickleshare               0.7.5           pyhd3eb1b0_1003  \n",
      "pip                       23.3.1           py38h06a4308_0  \n",
      "pkgutil-resolve-name      1.3.10           py38h06a4308_0  \n",
      "platformdirs              3.10.0           py38h06a4308_0  \n",
      "plotly                    5.9.0            py38h06a4308_0  \n",
      "prometheus_client         0.14.1           py38h06a4308_0  \n",
      "prompt-toolkit            3.0.36           py38h06a4308_0  \n",
      "psutil                    5.9.0            py38h5eee18b_0  \n",
      "ptyprocess                0.7.0              pyhd3eb1b0_2  \n",
      "pure_eval                 0.2.2              pyhd3eb1b0_0  \n",
      "py2bit                    0.3.0            py38he5da3d1_8    bioconda\n",
      "pybigwig                  0.3.22           py38h5df1436_1    bioconda\n",
      "pycparser                 2.21               pyhd3eb1b0_0  \n",
      "pyee                      8.2.2            py38h06a4308_0  \n",
      "pygments                  2.15.1           py38h06a4308_1  \n",
      "pyopenssl                 23.2.0           py38h06a4308_0  \n",
      "pyparsing                 3.0.9            py38h06a4308_0  \n",
      "pyppeteer                 1.0.2              pyhd8ed1ab_0    conda-forge\n",
      "pysam                     0.22.0           py38h15b938a_0    bioconda\n",
      "pysocks                   1.7.1            py38h06a4308_0  \n",
      "python                    3.8.18               h955ad1f_0  \n",
      "python-dateutil           2.8.2              pyhd3eb1b0_0  \n",
      "python-fastjsonschema     2.16.2           py38h06a4308_0  \n",
      "python-json-logger        2.0.7            py38h06a4308_0  \n",
      "python_abi                3.8                      2_cp38    conda-forge\n",
      "pytz                      2023.3.post1     py38h06a4308_0  \n",
      "pyyaml                    6.0.1            py38h5eee18b_0  \n",
      "pyzmq                     25.1.0           py38h6a678d5_0  \n",
      "readline                  8.2                  h5eee18b_0  \n",
      "referencing               0.30.2           py38h06a4308_0  \n",
      "requests                  2.31.0           py38h06a4308_0  \n",
      "rfc3339-validator         0.1.4            py38h06a4308_0  \n",
      "rfc3986-validator         0.1.1            py38h06a4308_0  \n",
      "rpds-py                   0.10.6           py38hb02cf49_0  \n",
      "scipy                     1.8.1            py38h1ee437e_0    conda-forge\n",
      "send2trash                1.8.2            py38h06a4308_0  \n",
      "setuptools                68.0.0           py38h06a4308_0  \n",
      "six                       1.16.0             pyhd3eb1b0_1  \n",
      "sniffio                   1.2.0            py38h06a4308_1  \n",
      "soupsieve                 2.5              py38h06a4308_0  \n",
      "sqlite                    3.41.2               h5eee18b_0  \n",
      "stack_data                0.2.0              pyhd3eb1b0_0  \n",
      "tenacity                  8.2.2            py38h06a4308_0  \n",
      "terminado                 0.17.1           py38h06a4308_0  \n",
      "tinycss2                  1.2.1            py38h06a4308_0  \n",
      "tk                        8.6.12               h1ccaba5_0  \n",
      "tomli                     2.0.1            py38h06a4308_0  \n",
      "tornado                   6.3.3            py38h5eee18b_0  \n",
      "tqdm                      4.65.0           py38hb070fc8_0  \n",
      "traitlets                 5.7.1            py38h06a4308_0  \n",
      "typing-extensions         4.7.1            py38h06a4308_0  \n",
      "typing_extensions         4.7.1            py38h06a4308_0  \n",
      "urllib3                   1.26.18          py38h06a4308_0  \n",
      "wcwidth                   0.2.5              pyhd3eb1b0_0  \n",
      "webencodings              0.5.1                    py38_1  \n",
      "websocket-client          0.58.0           py38h06a4308_4  \n",
      "websockets                10.4             py38h5eee18b_1  \n",
      "wheel                     0.41.2           py38h06a4308_0  \n",
      "xz                        5.4.5                h5eee18b_0  \n",
      "yaml                      0.2.5                h7b6447c_0  \n",
      "zeromq                    4.3.4                h2531618_0  \n",
      "zipp                      3.11.0           py38h06a4308_0  \n",
      "zlib                      1.2.13               hd590300_5    conda-forge\n"
     ]
    }
   ],
   "source": [
    "conda list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "&nbsp;\n",
    "\n",
    "Crucially, does the command `deeptools` give you a help message."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: deeptools [-h] [--version]\n",
      "\n",
      "deepTools is a suite of python tools particularly developed for the efficient analysis of\n",
      "high-throughput sequencing data, such as ChIP-seq, RNA-seq or MNase-seq.\n",
      "\n",
      "Each tool should be called by its own name as in the following example:\n",
      "\n",
      " $ bamCoverage -b reads.bam -o coverage.bw\n",
      "\n",
      "If you find deepTools useful for your research please cite as:\n",
      "\n",
      "Ramírez, Fidel, Devon P. Ryan, Björn Grüning, Vivek Bhardwaj, Fabian Kilpert,\n",
      "Andreas S. Richter, Steffen Heyne, Friederike Dündar,\n",
      "and Thomas Manke. 2016. \"deepTools2: A next Generation Web Server for Deep-Sequencing\n",
      "Data Analysis.\" Nucleic Acids Research, April. doi:10.1093/nar/gkw257.\n",
      "\n",
      "[ Tools for BAM and bigWig file processing ]\n",
      "    multiBamSummary         compute read coverages over bam files. Output used for plotCorrelation or plotPCA\n",
      "    multiBigwigSummary      extract scores from bigwig files. Output used for plotCorrelation or plotPCA\n",
      "    correctGCBias           corrects GC bias from bam file. Don't use it with ChIP data\n",
      "    bamCoverage             computes read coverage per bins or regions\n",
      "    bamCompare              computes log2 ratio and other operations of read coverage of two samples per bins or regions\n",
      "    bigwigCompare           computes log2 ratio and other operations from bigwig scores of two samples per bins or regions\n",
      "    computeMatrix           prepares the data from bigwig scores for plotting with plotHeatmap or plotProfile\n",
      "    alignmentSieve          filters BAM alignments according to specified parameters, optionally producing a BEDPE file\n",
      "\n",
      "[ Tools for QC ]\n",
      "    plotCorrelation         plots heatmaps or scatterplots of data correlation\n",
      "    plotPCA                 plots PCA\n",
      "    plotFingerprint         plots the distribution of enriched regions\n",
      "    bamPEFragmentSize       returns the read length and paired-end distance from a bam file\n",
      "    computeGCBias           computes and plots the GC bias of a sample\n",
      "    plotCoverage            plots a histogram of read coverage\n",
      "    estimateReadFiltering   estimates the number of reads that will be filtered from a BAM file or files given certain criteria\n",
      "\n",
      "[Heatmaps and summary plots]\n",
      "    plotHeatmap             plots one or multiple heatmaps of user selected regions over different genomic scores\n",
      "    plotProfile             plots the average profile of user selected regions over different genomic scores\n",
      "    plotEnrichment          plots the read/fragment coverage of one or more sets of regions\n",
      "\n",
      "[Miscellaneous]\n",
      "    computeMatrixOperations Modifies the output of computeMatrix in a variety of ways.\n",
      "\n",
      "For more information visit: http://deeptools.readthedocs.org\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help  show this help message and exit\n",
      "  --version   show program's version number and exit\n"
     ]
    }
   ],
   "source": [
    "deeptools"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "&nbsp;\n",
    "\n",
    "We should be good to go!!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: bigwigCompare [-h] [--version] [--binSize INT bp]\n",
      "                     [--region CHR:START:END]\n",
      "                     [--blackListFileName BED file [BED file ...]]\n",
      "                     [--numberOfProcessors INT] [--verbose] --outFileName\n",
      "                     FILENAME [--outFileFormat {bigwig,bedgraph}]\n",
      "                     [--deepBlueURL DEEPBLUEURL] [--userKey USERKEY]\n",
      "                     [--deepBlueTempDir DEEPBLUETEMPDIR] [--deepBlueKeepTemp]\n",
      "                     --bigwig1 Bigwig file --bigwig2 Bigwig file\n",
      "                     [--scaleFactors SCALEFACTORS]\n",
      "                     [--pseudocount PSEUDOCOUNT [PSEUDOCOUNT ...]]\n",
      "                     [--skipZeroOverZero]\n",
      "                     [--operation {log2,ratio,subtract,add,mean,reciprocal_ratio,first,second}]\n",
      "                     [--skipNonCoveredRegions]\n",
      "\n",
      "This tool compares two bigWig files based on the number of mapped reads. To\n",
      "compare the bigWig files, the genome is partitioned into bins of equal size,\n",
      "then the number of reads found in each BAM file are counted per bin and\n",
      "finally a summary value is reported. This value can be the ratio of the number\n",
      "of readsper bin, the log2 of the ratio, the sum or the difference.\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  --bigwig1 Bigwig file, -b1 Bigwig file\n",
      "                        Bigwig file 1. Usually the file for the treatment.\n",
      "                        (default: None)\n",
      "  --bigwig2 Bigwig file, -b2 Bigwig file\n",
      "                        Bigwig file 2. Usually the file for the control.\n",
      "                        (default: None)\n",
      "  --scaleFactors SCALEFACTORS\n",
      "                        Set this parameter to multipy the bigwig values by a\n",
      "                        constant. The format is scaleFactor1:scaleFactor2. For\n",
      "                        example 0.7:1 to scale the first bigwig file by 0.7\n",
      "                        while not scaling the second bigwig file (default:\n",
      "                        None)\n",
      "  --pseudocount PSEUDOCOUNT [PSEUDOCOUNT ...]\n",
      "                        A small number to avoid x/0. Only useful together with\n",
      "                        --operation log2 or --operation ratio. You can specify\n",
      "                        different values as pseudocounts for the numerator and\n",
      "                        the denominator by providing two values (the first\n",
      "                        value is used as the numerator pseudocount and the\n",
      "                        second the denominator pseudocount). (Default: 1)\n",
      "  --skipZeroOverZero    Skip bins where BOTH BAM files lack coverage. This is\n",
      "                        determined BEFORE any applicable pseudocount is added.\n",
      "                        (default: False)\n",
      "  --operation {log2,ratio,subtract,add,mean,reciprocal_ratio,first,second}\n",
      "                        The default is to output the log2ratio of the two\n",
      "                        samples. The reciprocal ratio returns the the negative\n",
      "                        of the inverse of the ratio if the ratio is less than\n",
      "                        0. The resulting values are interpreted as negative\n",
      "                        fold changes. Instead of performing a computation\n",
      "                        using both files, the scaled signal can alternatively\n",
      "                        be output for the first or second file using the '--\n",
      "                        operation first' or '--operation second' (Default:\n",
      "                        log2)\n",
      "  --skipNonCoveredRegions, --skipNAs\n",
      "                        This parameter determines if non-covered regions\n",
      "                        (regions without a score) in the bigWig files should\n",
      "                        be skipped. The default is to treat those regions as\n",
      "                        having a value of zero. The decision to skip non-\n",
      "                        covered regions depends on the interpretation of the\n",
      "                        data. Non-covered regions in a bigWig file may\n",
      "                        represent repetitive regions that should be skipped.\n",
      "                        Alternatively, the interpretation of non-covered\n",
      "                        regions as zeros may be wrong and this option should\n",
      "                        be used (default: False)\n",
      "\n",
      "Optional arguments:\n",
      "  --version             show program's version number and exit\n",
      "  --binSize INT bp, -bs INT bp\n",
      "                        Size of the bins, in bases, for the output of the\n",
      "                        bigwig/bedgraph file. (Default: 50)\n",
      "  --region CHR:START:END, -r CHR:START:END\n",
      "                        Region of the genome to limit the operation to - this\n",
      "                        is useful when testing parameters to reduce the\n",
      "                        computing time. The format is chr:start:end, for\n",
      "                        example --region chr10 or --region\n",
      "                        chr10:456700:891000. (default: None)\n",
      "  --blackListFileName BED file [BED file ...], -bl BED file [BED file ...]\n",
      "                        A BED or GTF file containing regions that should be\n",
      "                        excluded from all analyses. Currently this works by\n",
      "                        rejecting genomic chunks that happen to overlap an\n",
      "                        entry. Consequently, for BAM files, if a read\n",
      "                        partially overlaps a blacklisted region or a fragment\n",
      "                        spans over it, then the read/fragment might still be\n",
      "                        considered. Please note that you should adjust the\n",
      "                        effective genome size, if relevant. (default: None)\n",
      "  --numberOfProcessors INT, -p INT\n",
      "                        Number of processors to use. Type \"max/2\" to use half\n",
      "                        the maximum number of processors or \"max\" to use all\n",
      "                        available processors. (Default: 1)\n",
      "  --verbose, -v         Set to see processing messages. (default: False)\n",
      "\n",
      "Output:\n",
      "  --outFileName FILENAME, -o FILENAME\n",
      "                        Output file name. (default: None)\n",
      "  --outFileFormat {bigwig,bedgraph}, -of {bigwig,bedgraph}\n",
      "                        Output file type. Either \"bigwig\" or \"bedgraph\".\n",
      "                        (default: bigwig)\n",
      "\n",
      "deepBlue arguments:\n",
      "  Options used only for remote bedgraph/wig files hosted on deepBlue\n",
      "\n",
      "  --deepBlueURL DEEPBLUEURL\n",
      "                        For remote files bedgraph/wiggle files hosted on\n",
      "                        deepBlue, this specifies the server URL. The default\n",
      "                        is \"http://deepblue.mpi-inf.mpg.de/xmlrpc\", which\n",
      "                        should not be changed without good reason. (default:\n",
      "                        http://deepblue.mpi-inf.mpg.de/xmlrpc)\n",
      "  --userKey USERKEY     For remote files bedgraph/wiggle files hosted on\n",
      "                        deepBlue, this specifies the user key to use for\n",
      "                        access. The default is \"anonymous_key\", which suffices\n",
      "                        for public datasets. If you need access to a\n",
      "                        restricted access/private dataset, then request a key\n",
      "                        from deepBlue and specify it here. (default:\n",
      "                        anonymous_key)\n",
      "  --deepBlueTempDir DEEPBLUETEMPDIR\n",
      "                        If specified, temporary files from preloading datasets\n",
      "                        from deepBlue will be written here (note, this\n",
      "                        directory must exist). If not specified, where ever\n",
      "                        temporary files would normally be written on your\n",
      "                        system is used. (default: None)\n",
      "  --deepBlueKeepTemp    If specified, temporary bigWig files from preloading\n",
      "                        deepBlue datasets are not deleted. A message will be\n",
      "                        printed noting where these files are and what sample\n",
      "                        they correspond to. These can then be used if you wish\n",
      "                        to analyse the same sample with the same regions\n",
      "                        again. (default: False)\n"
     ]
    }
   ],
   "source": [
    "bigwigCompare "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
