---
title: "Loops and Conditions CM515"
author: "Me"
date: "2025-04-17"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Peak Percentages


I have determined a number of ChIP-seq peaks in putative enhancer regions and would like to know what percentage of these peaks are located on each chromosome. I have created code to calculate this, but it is becoming extremely tedious to manually type each line... especially since the yellow fever mosquito has over twenty thousand unassembled contigs! Can you modify my code below so that it uses a loop to print out the percentages for each unique chromosome in my dataframe?



```{r}
enhancer_peaks=read.csv("ExamplePeaksSheet.csv")

unique(enhancer_peaks$chromosome)

print(sum(enhancer_peaks$chromosome=="AaegL5_1")/nrow(enhancer_peaks))
print(sum(enhancer_peaks$chromosome=="AaegL5_2")/nrow(enhancer_peaks))
print(sum(enhancer_peaks$chromosome=="AaegL5_3")/nrow(enhancer_peaks))
print(sum(enhancer_peaks$chromosome=="NIGP01002131")/nrow(enhancer_peaks))



```

## Procuring Promoters

A number of important proteins (and histone modifications) occur primarily in the promoter region of genes and it is useful to have a list of these regions.
For this exercise, each of these genes is represented by a single line in a bed file (read into a dataframe below), where in which the first column represents the chromosome, the second the start coordinate,
and the third the end coordinate. Using a loop, return a dataframe that is identical except the start and end coordinates of
each row represent the coordinates from 2000 base pairs upstream of the gene to the start of the gene.

```{r}
aaegypti_genes=read.table("CM515.bed")

```


There is something we failed to consider in the above problem, which is that genes can be on either the forward (+) or reverse (-) strand of a chromosome. This is represented by the 6th column of our gff. 

Use an if/else statement to modify your above loop so that promoters for genes on the reverse strand represent the 2000 bp after the original end coordinate of the gene.

```{r}
aaegypti_genes=read.table("CM515.bed")

```